### PR TITLE
DCG edits

### PR DESCRIPTION
**Pull request by GodloveD**
I'm basically just going through all of the user docs from start to finish trying to find things that are outdated or incorrect and also attempting to edit for clarity while I'm at it.  

This is not ready for merge yet.  I've only gotten as far as the recipe file section.  But I wanted to push so folks can view changes and comment if they like.  

Still so much work to be done...

### CODE REVIEW

**Comment by GodloveD** [comment]
Hey @vsoch.  Would you mind if we go ahead and merge this?  @bauerm97 and @gmkurtzer are also writing docs and they want to make sure they are working with the latest greatest version.

**Comment by GodloveD** [comment]
Oh wait.  Just saw there were conflicts.  Let me have a look.

**Comment by GodloveD** [comment]
OK should be good to go now.  @vsoch can you take a look when able please?

**Comment by vsoch** [comment]
yep looking now

File: _data/sidebars/user_docs.yml [hunk 1]
```diff
@@ -23,14 +23,19 @@ entries:
       url: /quickstart
       output: web, pdf
 
-    - title: Build an Image
+    - title: Build a Container
       url: /docs-build
```
Comment (vsoch) [inline _data/sidebars/user_docs.yml:27]: +1. I suspect in 3.0 this will be "Build a Robot"

File: _data/sidebars/user_docs.yml [hunk 2]
```diff
@@ -23,14 +23,19 @@ entries:
...
     - title: Container Recipes
       url: /docs-recipes
       output: web, pdf
 
+
+    - title: Singularity Flow
+      url: /docs-flow
+      output: web, pdf
+
     - title: Singularity Flow
```
Comment (vsoch) [inline _data/sidebars/user_docs.yml]: Why is this here twice?

Comment (vsoch) [reply _data/sidebars/user_docs.yml]: Can we remove one?

File: pages/docs/user-docs/action-build.md [hunk 3]
```diff
@@ -1,303 +1,185 @@
...
-title: Build an Image
+title: Build a Container
 sidebar: user_docs
 permalink: docs-build
 folder: docs
+toc: false
 ---
 
-Build is the process where we install an operating system and then configure it appropriately for a specified need. To do this we use a [Singularity recipe](/docs-recipes) file (a text file called `Singularity`) which is a recipe of how to specifically build the container. Here we will overview the usage, best practices, and quick examples.
+The `build` command lets you obtain a container.
```
Comment (vsoch) [inline pages/docs/user-docs/action-build.md]: I would just delete "The build command lets you obtain  container." Obtain is more similar to pull, and building is what it is... building. It's a cleaner, strong start to just begin with "Build" is the "Swiss..."

File: pages/docs/user-docs/action-build.md [hunk 4]
```diff
@@ -1,303 +1,185 @@
...
 ---
 
-Build is the process where we install an operating system and then configure it appropriately for a specified need. To do this we use a [Singularity recipe](/docs-recipes) file (a text file called `Singularity`) which is a recipe of how to specifically build the container. Here we will overview the usage, best practices, and quick examples.
+The `build` command lets you obtain a container. 
 
+`build` is the "Swiss army knife" of container creation.  You can use it to download and assemble pre-built containers from external resources like [Singularity Hub](https://singularity-hub.org/) and [Docker Hub](https://hub.docker.com/).  You can use it to convert containers between the various formats supported by Singularity.  And you can use it in conjunction with a [Singularity Recipe](/docs-recipes) file to create a container from scratch and customized it to fit your needs. 
 
-**Getting Started**
-For a detailed **walkhrough of getting started** with build, we will point you to the <a href="/quickstart">quickstart</a>. 
+{% include toc.html %}
```
Comment (vsoch) [inline pages/docs/user-docs/action-build.md]: +1 here. If the user is on the build page, they probably just came from Quickstart, previous in navigation.

File: pages/docs/user-docs/action-build.md [hunk 5]
```diff
@@ -1,303 +1,185 @@
...
 
-**Build Environment**
-If you want to customize the cache, specify Docker credentials, or any custom tweaks to your build environment, see <a href="/build-environment">build environment</a>. 
+## Overview
```
Comment (vsoch) [inline pages/docs/user-docs/action-build.md]: This needs to stay - all of these variables are essential to know, most relevant for build, and it's not a big enough section to warrant it's own thing. Let's add this guy back!

**GodloveD sez** I don't know how to reply to this so I'm just editing your comment.  I just moved this to the bottom of the page instead.  It's still there.

File: pages/docs/user-docs/action-build.md [hunk 6]
```diff
@@ -1,303 +1,185 @@
...
+- URI beginning with **docker://** to build from Docker Hub
+- path to a **existing container** on your local machine
+- path to a **directory** to build from a sandbox
+- path to an **archive** in .tar or compressed .tar.gz format
+- path to a **[Recipe file](docs-recipes)**
 
-{% include toc.html %}
+In addition `build` can produce containers in three different formats.  Formats types can be specified by passing the following options to build.  
 
-## Bases
```
Comment (vsoch) [inline pages/docs/user-docs/action-build.md]: Can we still have a header for this? Eg `## Container Formats` so the user could jump to it quickly?

File: pages/docs/user-docs/action-build.md [hunk 7]
```diff
@@ -1,303 +1,185 @@
...
-You have many options for the base of your build. See the linked page for each for more details, or read the [Usage](#usage) section below for a summary.
+- compressed read-only **squashfs** file system suitable for production (default)
+- writable **ext3** file system suitable for interactive development (`--writable` option)
+- writable **(ch)root directory** called a sandbox for interactive development (`--sandbox` option)
 
-### Singularity Recipes
-For a reproducible container, the recommended practice is to build by way of a Singularity recipe file. This also makes it easy to add files, environment variables, and install custom software, and still start from your base of choice (e.g., Docker). The absolute minimum required for a recipe is a base, and here are your options:
+Because `build` can accept an existing container as a target and create a container in any of these three formats you can convert existing containers from one format to another.
 
+Some things need root...
```
Comment (vsoch) [inline pages/docs/user-docs/action-build.md]: I don't understand this bit... what is the ... ?

File: pages/docs/user-docs/action-build.md [hunk 8]
```diff
@@ -1,303 +1,185 @@
...
 
-**Singularity Hub**
-```
-Bootstrap: shub
-From: vsoch/hello-world
-```
+## Downloading a pre-built container from Singularity Hub
+You can use the `build` command to download a container from Singularity Hub. 
 
-[**Docker**](/docs-docker)
```
Comment (vsoch) [inline pages/docs/user-docs/action-build.md]: This still doesn't make sense to me. If you are using build with singularity hub, then you are creating a new image FROM one on singularity hub. If that is incorrect, then it should be  a pull. But saying "download container" and then saying we are building is really confusing for the user.

File: pages/docs/user-docs/action-build.md [hunk 9]
```diff
@@ -1,303 +1,185 @@
...
-MirrorURL: http://us.archive.ubuntu.com/ubuntu/
-```
+But default the container will be converted to a compressed, read-only squashfs file.  If you want your container in a different format use the `--writable` or `--sandbox` options.
 
-[**Self**](/build-self)
-```
-Bootstrap: self
-```
+## Downloading a pre-built container from Docker Hub
+You can use `build` to download layers from Docker Hub and assemble them into Singularity containers.
```
Comment (vsoch) [inline pages/docs/user-docs/action-build.md]: This is still misleading. The container is not pre built. It is layers from Docker Hub that are assembled at build time to make a container. These subtle details are important and we need to be clear in the choice of words.

File: pages/docs/user-docs/action-build.md [hunk 10]
```diff
@@ -1,303 +1,185 @@
...
 
-You can think of the different options as follows:
+## Converting containers from one format to another
+If you already have a container saved locally, you can use it as a target to build a new container. This allows you convert containers from one format to another.  For example if you had a squashfs container called `production.simg` and wanted to convert it to a writeable ext3 container called `development.img` you could:
 
-**create writable development folder**
 ```
-sudo singularity build --writable container/ Singularity
+$ singularity build --writable development.img production.simg
 ```
```
Comment (vsoch) [inline pages/docs/user-docs/action-build.md:99]: very good choice of container names here! :)

File: pages/docs/user-docs/action-build.md [hunk 11]
```diff
@@ -1,303 +1,185 @@
...
-    -l|--low       Specify low threshold (all checks, default) 
-    -m|--med       Perform medium and high checks
-    -h|--high      Perform only checks at level high
-
-See singularity check --help for available tags
+$ singularty build production2 development/
 ```
+You can convert the three supported container formats using any combination.
 
+Use care when converting writable ext3 images or sandbox directories to the default squashfs file format.  If changes were made to the writable container before conversion, there is no record of those changes in the Recipe file rendering your container non-reproducible.  It is a best practice to build your immutable production containers directly from a Recipe file instead.
```
Comment (vsoch) [inline pages/docs/user-docs/action-build.md]: why is Recipe capitalized?

File: pages/docs/user-docs/action-build.md [hunk 12]
```diff
@@ -1,303 +1,185 @@
...
 
-## Examples
-Let's quickly look at some definition file examples. These come straight from `singularity help build`. First, here
-is a recipe for build that serves only to describe the sections.
+## Building containers from Recipe files
+Of course, Recipe files can be used as the target when building a container.  For detailed information on writing Recipe files, please see [the Container Recipes docs](docs-recipes).
```
Comment (vsoch) [inline pages/docs/user-docs/action-build.md]: The standard we've been using, in many places, Singularity Hub, docs, is "Singularity recipe" or build recipe.

File: pages/docs/user-docs/action-build.md [hunk 13]
```diff
@@ -1,303 +1,185 @@
...
+From: ubuntu:16.04
 
-Now let's look at a recipe that defines different <a href="https://containers-ftw.github.io/apps/" target="_blank">SCI-F apps</a>
+%post
+    apt-get -y update
+    apt-get -y install fortune cowsay lolcat
 
+%environment
+    export LC_ALL=C
+    export PATH=/usr/games:$PATH
```
Comment (vsoch) [inline pages/docs/user-docs/action-build.md:132]:
```
LC_ALL=C
PATH=/usr/games:$PATH
export LC_ALL PATH
```

File: pages/docs/user-docs/docs-environment-metadata.md [hunk 14]
```diff
@@ -6,36 +6,40 @@ folder: docs
...
 
-Singularity containers support environment variables and labels that can be added by user during the bootstrap process.
+Singularity containers support environment variables and labels that you can add to your container during the build process.
 
 {% include toc.html %}
 
 ## Environment
 
-If you are importing a Docker container, the environment will be imported as well. If you want to define custom environment variables in your bootstrap recipe file `Singularity` you can do that like this
+If you build a container from Singularity Hub or Docker Hub, the environment will be imported into the container at build time. You can also define custom environment variables in your Recipe file like so:
```
Comment (vsoch) [inline pages/docs/user-docs/docs-environment-metadata.md]: This doesn't make sense to me. Singularity Hub containers already have an environment, how is it imported? You are just pulling and using the previous container.

File: pages/docs/user-docs/docs-environment-metadata.md [hunk 15]
```diff
@@ -6,36 +6,40 @@ folder: docs
...
 
-```bash
-Bootstrap:docker
-From: ubuntu:latest
+```
+Bootstrap: shub
+From: singularityhub/ubuntu
```
Comment (vsoch) [inline pages/docs/user-docs/docs-environment-metadata.md]: I wouldn't use the singularityhub user as an example - it can be confusing for the user, they might think that any container should be prefixed with "singularityhub" use vsoch/hello-world instead.

File: pages/docs/user-docs/docs-environment-metadata.md [hunk 16]
```diff
@@ -6,36 +6,40 @@ folder: docs
...
 
 %environment
     VARIABLE_NAME=VARIABLE_VALUE
     export VARIABLE_NAME
 ```
 
-If you need to add an environment variable to your container during the `%post` section, you can do so using the `$SINGULARITY_ENVIRONMENT` variable with the following syntax:
+You may need to add environment variables to your container during the `$post` section.  For instance, maybe you will not know the appropriate value of a variable until you have installed some software.  
+
```
Comment (vsoch) [inline pages/docs/user-docs/docs-environment-metadata.md]: `%post`

File: pages/docs/user-docs/docs-recipes.md [hunk 17]
```diff
@@ -2,60 +2,67 @@
...
 
 ### Header
-The header is at the top of the file, and tells Singularity the base to use, and which one. Specifically:
+The header is at the top of the file, and tells Singularity the base Operating System that it should use to build the container.  It is composed of several keywords. Specifically:
 
- - **Bootstrap**: references the kind of base you want to use (e.g., docker, debootstrap, shub). For example, a shub bootstrap will pull containers for shub as bases. A Docker bootstrap will pull docker layers to start your image. For a full list see <a href="/build">build</a>
- - **From**: is the named container (shub) or reference to layers (Docker) that you want to use (e.g., vsoch/hello-world)
+ - `Bootstrap:` references the kind of base you want to use (e.g., docker, debootstrap, shub). For example, a shub bootstrap will pull containers for shub as bases. A Docker bootstrap will pull docker layers to start your image. For a full list see <a href="/build">build</a>
```
Comment (vsoch) [inline pages/docs/user-docs/docs-recipes.md]: This url should be /docs-build

File: pages/docs/user-docs/docs-recipes.md [hunk 18]
```diff
@@ -2,60 +2,67 @@
...
-For example, a very minimal Docker build might look like this:
+For example, a very minimal Singularity Hub build might look like this:
 
-```bash
-Bootstrap: docker
-From: ubuntu:latest
+```
+Bootstrap: shub
+From: singularityhub:ubuntu
 ```
```
Comment (vsoch) [inline pages/docs/user-docs/docs-recipes.md]: vsoch/hello-world

File: pages/docs/user-docs/docs-recipes.md [hunk 19]
```diff
@@ -2,60 +2,67 @@
...
 
-Each different kind of build base comes with particular details during build time, and you can read about them at the following links:
+Each build base requires particular details during build time.  You can read about them and see examples at the following links:
 
- - [yum base](/build-yum)
- - [debootstrap base](/build-debootstrap)
- - [arch linux base](/build-arch)
- - [yum base](/build-docker)
- - [local image base](/build-localimage)
+ - [shub](/build-docker) (images hosted on Singularity Hub)
```
Comment (vsoch) [inline pages/docs/user-docs/docs-recipes.md]: this url doesn't exist I don't think...

File: _data/sidebars/user_docs.yml (hunk 2 above)
Comment (GodloveD) [reply _data/sidebars/user_docs.yml]: Oh crap.  That was probably an error when I did the merge.  I'll look again.

File: pages/docs/user-docs/docs-recipes.md (hunk 19 above)
Comment (vsoch) [reply pages/docs/user-docs/docs-recipes.md]: (note both are for docker)

**Comment by vsoch** [comment]
Good start! Let met know when you are ready for second look.

File: pages/docs/user-docs/action-build.md (hunk 4 above)
Comment (GodloveD) [reply pages/docs/user-docs/action-build.md]: Yes my thoughts too.  And I think quickstart should be broad strokes and details should be in other user pages.  In other words we should be referring users to docs from quick start not referring users to quickstart from  docs.

**Comment by GodloveD** [comment]
@vsoch ready.  😸

**Review by vsoch — CHANGES_REQUESTED** [review]

**Comment by vsoch** [comment]
I'm guessing you didn't see all the comments above yet- I forgot to submit my "review" lol

File: pages/docs/user-docs/docs-environment-metadata.md (hunk 15 above)
Comment (GodloveD) [reply pages/docs/user-docs/docs-environment-metadata.md]: vsoch is less confusing than singularityhub? lol
We need a user named official or something.

File: pages/docs/user-docs/docs-environment-metadata.md (hunk 14 above)
Comment (GodloveD) [reply pages/docs/user-docs/docs-environment-metadata.md]: How about "the environment will be included with the container at build time" to cover both instances.

File: pages/docs/user-docs/action-build.md (hunk 11 above)
Comment (vsoch) [reply pages/docs/user-docs/action-build.md]: I would say Singularity recipe or build recipe

File: pages/docs/user-docs/action-build.md (hunk 9 above)
Comment (GodloveD) [reply pages/docs/user-docs/action-build.md]: How about download an "existing" container?

File: pages/docs/user-docs/action-build.md (hunk 10 above)
Comment (GodloveD) [reply pages/docs/user-docs/action-build.md:99]: thx

File: pages/docs/user-docs/action-build.md (hunk 11 above)
Comment (GodloveD) [reply pages/docs/user-docs/action-build.md]: Same reason Dockerfile is capitalized in their docs.  https://docs.docker.com/compose/gettingstarted/ .  Whatever we end up calling this thing (I actually prefer Definition file with the suffix .def) I think it should be a proper noun.

File: pages/docs/user-docs/action-build.md (hunk 12 above)
Comment (GodloveD) [reply pages/docs/user-docs/action-build.md]: The standard we've been using in the help is Definition file.  And when I give talks and teach workshops that's what I've been calling it too.  I understand that you like recipe now but I didn't realize it had become a standard. lol

File: pages/docs/user-docs/action-build.md (hunk 13 above)
Comment (GodloveD) [reply pages/docs/user-docs/action-build.md:132]: I think you are missing the point here.  There is no need for users to avoid bashisms in definition files.  It's not a best practice.  It's just pedantics that we enforce in the code so that Singularity can run on things like busybox.  I don't really want to start rewriting def file examples with syntax that many users would find awkward for no good reason.

File: pages/docs/user-docs/docs-environment-metadata.md (hunk 14 above)
Comment (GodloveD) [reply pages/docs/user-docs/docs-environment-metadata.md]: How about "the environment will be included with the container at build time" to cover both instances.

File: pages/docs/user-docs/docs-environment-metadata.md (hunk 16 above)
Comment (GodloveD) [reply pages/docs/user-docs/docs-environment-metadata.md]: thx

File: pages/docs/user-docs/docs-recipes.md (hunk 17 above)
Comment (GodloveD) [reply pages/docs/user-docs/docs-recipes.md]: thx

File: pages/docs/user-docs/docs-recipes.md (hunk 18 above)
Comment (GodloveD) [reply pages/docs/user-docs/docs-recipes.md]: I'm not sold on that.  What's the point of having the singularityhub user with the "official" containers if you don't encourage users to use them?

File: pages/docs/user-docs/docs-recipes.md (hunk 19 above)
Comment (GodloveD) [reply pages/docs/user-docs/docs-recipes.md]: Most of these urls don't exist.  And they're not going to until somebody can get a chance to write them lol.  This is a big job and is not going to get finished in a single PR.

File: pages/docs/user-docs/action-build.md (hunk 11 above)
Comment (vsoch) [reply pages/docs/user-docs/action-build.md]: @GodloveD it's been called Singularity (akin to Dockerfile) since early on. take a look at how Singularity Hub works - it finds a file called Singularity. if you move around the docs, and in all the code, it's Singularity. So you got the right idea! Just wrong word ::P

Comment (vsoch) [reply pages/docs/user-docs/action-build.md]: so here it would be appropriate to change to Singularity recipe or just build recipe.

File: pages/docs/user-docs/action-build.md (hunk 12 above)
Comment (vsoch) [reply pages/docs/user-docs/action-build.md]: I guess you haven't used Singularity Hub. I't's also, erm, published!

image
http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0177459

File: pages/docs/user-docs/action-build.md (hunk 13 above)
Comment (vsoch) [reply pages/docs/user-docs/action-build.md:132]: okay, that's ok with me then. 2 lines > 3!

File: pages/docs/user-docs/docs-environment-metadata.md (hunk 15 above)
Comment (vsoch) [reply pages/docs/user-docs/docs-environment-metadata.md]: we really just need a container that is guaranteed to persist and be updated over time. This is the image I do this with, so it makes sense. In the case that the github organization singularityhub goes away, that uri won't persist. I also built, manage, everything for singularity hub, so it's official yo!

File: pages/docs/user-docs/docs-recipes.md (hunk 18 above)
Comment (vsoch) [reply pages/docs/user-docs/docs-recipes.md]: I don't actually think it's useful. They aren't being used.

File: pages/docs/user-docs/action-build.md [hunk 20]
```diff
@@ -111,12 +111,12 @@ $ singularty build production2 development/
...
+Use care when converting writable ext3 images or sandbox directories to the default squashfs file format.  If changes were made to the writable container before conversion, there is no record of those changes in the Singularity recipe file rendering your container non-reproducible.  It is a best practice to build your immutable production containers directly from a Singularity recipe file instead.
 
-## Building containers from Recipe files
-Of course, Recipe files can be used as the target when building a container.  For detailed information on writing Recipe files, please see [the Container Recipes docs](docs-recipes).
+## Building containers from Singularit recipe files
+Of course, Singularity recipe files can be used as the target when building a container.  For detailed information on writing Singularity recipe files, please see [the Container Recipes docs](docs-recipes).
```
Comment (vsoch) [inline pages/docs/user-docs/action-build.md]: the "y" ran away!

File: pages/docs/user-docs/docs-environment-metadata.md (hunk 14 above)
Comment (vsoch) [reply pages/docs/user-docs/docs-environment-metadata.md]: that would work!

**Review by vsoch — APPROVED** [review]

**Comment by vsoch** [comment]
Totally fantastic work! And I just figured out that I can make small tweaks directly to the PR, so hopefully can help with some of the "oh lord @vsoch that's too nit picky detailed!" for next time. I really appreciate the time and care to dot all the i's and cross the t's. Merge away!

**Comment by GodloveD** [comment]
Thank you!  😸

**Comment by vsoch** [comment]
:D
